sympy/polys/rings.py

```diff
@@ -2958,6 +3028,45 @@ def subresultants(self, g, x=None):
 
         return R
 
+    def free_variable(self):
+        """
+        Examples
+        ========
+
+        >>> from sympy import ZZ, ring
+        >>> R, x, y, z = ring("x, y, z", ZZ)
+        >>> p = 3*x**2 + 2*y**2 + 5*z**3 + 7*x**2*y
+        >>> p.free_variable()
+        {0, 1, 2}
+
+        """
+        p = self
+        exponents = list(map(sum, zip(*p)))
+        return {n for n, e in enumerate(exponents) if e}
+
+    def main_variable(self):
+        """
+        Find the smallest index of the free symbolic variable in a given
+        polynomial.
+
+        Examples
+        ========
+
+        >>> from sympy import ZZ, ring
+        >>> R, x, y = ring("x, y", ZZ)
+        >>> p = y**2 + 2 - y**3 + 4*y
+        >>> p.main_variable()
+        1
+
+        >>> p = x*y**2 - y**3 + 4*y
+        >>> p.main_variable()
+        0
+
+        """
+        p = self
+        free_sym = p.free_variable()
+        return min(free_sym)
```

Comment:
This is another example where we can have an early return: if any monomial includes the generator with index `0` then we do not need to loop over the rest. It is very likely that either the first or the second monomial (if the first is the zero monomial) would involve the first generator so this is worth doing.

I would make this something like:
```
for monomial in self.itermonomials():
    leading_sym = ...
    if leading_sym == 0:
        return leading_sym
return leading_sym
```